README.md
# Elephant Scale Sandbox

## What is it?
This is a virtual machine sandbox image to practice and learn Big Data and Data Science applications.

## Why?
Running Big Data applications (Spark / Cassandra / Hadoop) can be a little convoluted because of all the dependencies.  This can be even more of a  hassle in Windows.  We hope this VM Sandbox will make things easier.

## Provided By Elephant Scale

[Elephant Scale](http://elephantscale.com) teaches [Big Data & AI / Data Science classes](http://elephantscale.com/training/).   This sandbox is a replica of our virtualized environment.  

#### Looking for training?
[Checkout our training classes in Big Data and Data Science](http://elephantscale.com/training/)

[Read more about the Sanxbox](http://elephantscale.com/sandbox/)


## Where to get it?
Currently OVA based virtual machine image is available.  Docker images coming 'soon'.  
Note : These are **LARGE** downloads (10G+ in size).  Download when you have good bandwidth.

- Latest version : v6
- Release date : 2017-11-12
- [Download link](https://s3.amazonaws.com/elephantscale-public/sandbox/ES_Sandbox_v6.ova)
- For older versions see [changelog](changelog.md)

## How to run it?
- You need a virtual machine 'player'.  Any of these would work:
  - [Virtual Box](https://www.virtualbox.org/wiki/VirtualBox)
  - [VMWare Player](http://www.vmware.com/products/player/playerpro-evaluation.html)
- Download the latest sandbox image
- Double click on the 'OVA' file open it.

## Access
Login : **student**  
password : **bigdata123**  

See [intro lab](labs/intro.md) for a screencast.

**Connectivity:**
- Use VM GUI : when you open this OVA file in a VM environment you will be logged into the Ubuntu desktop
- SSH via port 22
- from host machine
```
    $   ssh -l student  -p 2222 localhost
```

## What can I run?
This VM is tested with following Big Data stack.
- Spark v2.x and v1.6
- BigDL 0.3+
- Cassandra v3.x
- Kafka v0.10
- Storm v1.x
- Zookeeper v3.4.8


## Labs
If you are enrolled in our classes, you will get a lab bundle. Also you can run any open source labs as well.

### Tutorials
1. [Introduction](tutorials/intro.md)
2. [Running Spark](tutorials/running-spark.md)
3. [Running Cassandra](tutorials/running-cassandra.md)
4. [Choosing a Python version](tutorials/python-versions.md)
5. [Running BigDL](tutorials/running-bigdl.md)

Checkout our [Sandbox channel](https://vimeo.com/channels/1237347) for more videos.

## Software Installed
- Based on [Ubuntu 16.04 LTS](http://ubuntu.org)
- Most software is in **/usr/local/apps**  (also **~/apps**)
- Java / Scala
    - [Oracle JDK v1.8.0_u131](http://www.oracle.com/technetwork/java/javase/downloads/index-jsp-138363.html)
    - [Maven v3.5.0](https://maven.apache.org/)
    - [Scala v2.12.2](http://scala-lang.org/)
    - [SBT v0.13.15](http://www.scala-sbt.org/)
- Metrics
    - [Graphite](https://graphiteapp.org/)
    - [Grafana](https://grafana.com/)
- Python
    - Python 3.6
    - [Anaconda v4.3.1](https://www.continuum.io/anaconda-overview)
    - [Jupyter](http://jupyter.org/)
- Editors :
    - [Sublime 3](https://www.sublimetext.com/)
    - [Atom](https://atom.io/)
- IDEs
    - [Eclipse Neon](https://eclipse.org/)  - ~/apps/eclipse/java-neon/eclipse/eclipse
    - [IntelliJ Community Edition ](https://www.jetbrains.com/idea/) - ~/apps/idea/bin/idea.sh
- Big Data applications supported:
    - Spark
    - BigDL
    - Cassandra
    - Kafka
    - Storm
    - Zookeeper



## Changelog
See version history in [changelog](changelog.md)

## Feedback
We welcome your feedback about the sandbox.
- send an email to info@elephantscale.com
- or open a issue at the [Github page](https://github.com/elephantscale/sandbox)


changelog.md
# Changelog

To take a snapshot:
     ~/cleanup-before-imaging.sh
     and take a snapshot

Here is the Changelog in reverse order.

## Version : V6,  Date : 2017-11-12
[Download link](https://s3.amazonaws.com/elephantscale-public/sandbox/ES_Sandbox_v6.ova)
- installed Tensorflow, Tensorboard

## Version : V5,  Date : 2017-11-10
[Download link](https://s3.amazonaws.com/elephantscale-public/sandbox/ES_Sandbox_V5.ova)
- installed BigDL

## Version : V4,  Date : 2017-07-14
[Download link](https://s3.amazonaws.com/elephantscale-public/sandbox/ES_Sandbox_V4.ova)
- installed Graphite & Grafana

## Version : V3,  Date : 2017-05-02
[Download link](https://s3.amazonaws.com/elephantscale-public/sandbox/ES_Sandbox_V3.ova)
- update system
```
  $  sudo apt update;   sudo apt upgrade -y
```

- making Python3 as default python.  Use `python3` and `pip3`  run commands.  
To use Anaconda as default python, uncomment the line in  `~/.bashrc`
```
    #export PATH=/usr/local/apps/anaconda/bin:$PATH
```

- installed python libs
```bash
  $  sudo apt install -y python3-pip  python3-dev  python3-numpy
```

- installed Jupyter
```bash
  $   sudo pip3 install jupyter
```

- installed Tensorflow libraries
```bash
  $  sudo pip3 install tensorflow
```

## Version : V2,  Date : 2017-04-24
[Download link](https://s3.amazonaws.com/elephantscale-public/sandbox/ES_Sandbox_V2.ova)
- expanding disk to 100G
        $   VBoxManage clonehd  ES_Sandbox_V1-disk1.vmdk   disk1.vdi  --format vdi
        $   VBoxManage modifyhd disk1.vdi  --resize 100000
    add this as disk1 (using VirtualBox settings)
    when the machine boots up, use gparted  to resize partition

- adding data dir
    $  sudo mkdir /data
    $  sudo chown $USER /data
    $  ln -s /data  ~/data
- uploading some data


## version : V1,  Date : 2017-04-21
[Download link](https://s3.amazonaws.com/elephantscale-public/sandbox/ES_Sandbox_V1.ova)

- increase CPU --> 2  
increase memory --> 4096 M

- changed user name : student

- software install dir:  /usr/local/apps  
has a pointer :  /home/student/apps --> /usr/local/apps  
$   sudo   chown $USER  /usr/local/apps

- installed the following in /usr/local/apps  
    sun jdk 1.8,  scala, mvn ,  sublime, atom, anaconda, sbt, anaconda

- updated ~/.bashrc  
    see at the bottom

- Disabled sudo password  
added the following.  
$  sudo visudo

        username ALL=(ALL) NOPASSWD: ALL

- installing
```bash
    $    sudo apt install -y   vim  vim-gtk  git  atop  jq  iftop  strace sysstat
```

- install SSH  
```
    $  sudo apt install -y ssh
```

- Added port forwarding for SSH from localhost 2222   (Network --> Advanced --> Port Forwarding)
```
  ssh -p 2222 -l student  localhost
```

- Expanded virtual disk to 50G. Added a swap disk

- installed Anaconda python distribution in /usr/local/apps/anaconda.
This is the default python

- installed lein in /usr/local/bin
```
     wget https://raw.githubusercontent.com/technomancy/leiningen/stable/bin/lein
     sudo mv lein /usr/local/bin
     sudo chmod 755 /usr/local/bin/lein
```

- downloaded Big Data stuff into  ~/files  or ~/Downloads   


data/twinkle/create-data-files.sh
#!/bin/bash

## creates large *text* files (dd creates binary files)
## usage :
##    1) modify the number of iterations (in for loop, around 24 will
##       give you 1G data)
##    2) run as :  ./create-data-files.sh
##       this will create a bunch of *.data files with various sizes


if [ ! -e a.data ]
then
        echo -n "creating master data file..."
        cat sample.txt > a.data
        rm -f a1.data
        for i in {1..24}
        do
                echo -n "$i..."
                cat a.data   a.data >> a1.data
                mv a1.data a.data
                #du -skh a.data
        done
        echo "done"
fi

## which block size to use?
BS=1048576
#BS=1000000

## creates multiple sized files
echo "creating 1M.data..."
dd  if=a.data  of=1M.data  bs=$BS   count=1   #  1 M file

echo "creating 10M.data..."
dd  if=a.data  of=10M.data  bs=$BS   count=10 #  10 M file

echo "creating 100M.data..."
dd  if=a.data  of=100M.data  bs=$BS   count=100 #  100 M file

echo "creating 200M.data..."
dd  if=a.data  of=200M.data  bs=$BS   count=200 #  200 M file

echo "creating 500M.data..."
dd  if=a.data  of=500M.data  bs=$BS   count=500 #  500 M file

echo "creating 1G.data..."
dd  if=a.data  of=1G.data  bs=$BS   count=1000 #  1G file

data/clickstream/gen-clickstream.py
#!/usr/bin/env python

## generates mock click logs data
## log format
##   timestamp (in ms), ip_address, user, action, domain, campaign_id, cost, session_id

## timestamp converstions testing site : http://www.epochconverter.com/


## ----- config
days=10
entries_per_day=100000
# file size calculations
#  format     entries_per_day     file_size in MB
#  csv       10000000             830 M file
#  json      1000000              169 M
#  json      3200000              500 M

log_format="csv"
#log_format="json"
## --- end config


import os
import datetime as dt
import random
import json

domains = ['facebook.com',  'yahoo.com',   'google.com',   'zynga.com',    'wikipedia.org',   'sf.craigslist.org',   'twitter.com',    'amazon.com',    'flickr.com',    'cnn.com',      'usatoday.com',      'npr.org',    'foxnews.com',    'comedycentral.com',   'youtube.com',   'hulu.com',   'bbc.co.uk',  'nytimes.com',   'sfgate.com',   'funnyordie.com']

actions = ['viewed', 'clicked', 'blocked']
total_users = 100000
total_ips = 1000
total_sessions = 100
total_campaigns = 20

# overwrite this function to customize log generation
def generate_log(timestamp):
  user = "user_%d" % random.randint(1,total_users)
  action = random.choice(actions)
  domain = random.choice(domains)
  ip_address = "ip_%d" % random.randint(1,total_ips)
  campaign = "campaign_%d" % random.randint(1,total_campaigns)
  session = "session_%s" % random.randint(1,total_sessions)
  #cost is in cents, could be zero
  cost = random.randint(1,200) - 20
  if (cost < 0):
    cost = 0

  #csv
  if (log_format == 'csv'):
    logline = "%s,%s,%s,%s,%s,%s,%s,%s" % (timestamp, ip_address, user, action, domain, campaign, cost, session)

  # generate JSON format
  if (log_format == 'json'):
    dict={'timestamp': timestamp, 'ip': ip_address, 'user': user, 'action': action, 'domain': domain,  'campaign':campaign, 'cost': cost,  'session': session}
    logline = json.dumps(dict)


  #print logline
  return logline



#main
## --- script main
if __name__ == '__main__':
  time_inc_ms = int ((24.0*3600*1000)/entries_per_day)
  #print "time inc ms", time_inc_ms
  #epoch = dt.datetime.fromtimestamp(0)
  epoch = dt.datetime(1970,1,1)

  year_start = dt.datetime(2015, 1, 1)
  for day in range(0, days):
    day_delta = dt.timedelta(days=day)
    start_ts = year_start + day_delta
    #end_ts = dt.datetime(start_ts.year, start_ts.month, start_ts.day, 23, 59, 59)
    end_ts = dt.datetime(start_ts.year, start_ts.month, start_ts.day+1, 0, 0, 0)
    filename = start_ts.strftime("%Y-%m-%d") 
    if (log_format == 'csv'):
      filename = filename + ".csv"
    if (log_format == 'json'):
      filename = filename + ".json"

    #print start_ts
    #print end_ts
    last_ts = start_ts

    with open(filename, "w") as fout:
      print "generating log ", filename
      while (last_ts < end_ts):
        delta_since_epoch = last_ts - epoch
        millis = int((delta_since_epoch.microseconds + (delta_since_epoch.seconds + delta_since_epoch.days * 24 * 3600) * 10**6) / 1e3)
        #print "last ts", last_ts
        #print "millis",  millis
        logline = generate_log(millis)
        fout.write(logline + "\n")

        last_ts = last_ts + dt.timedelta(milliseconds=time_inc_ms)



data/clickstream/gen-clickstream-json.py
#!/usr/bin/env python

## generates mock click logs data
## log format
##   timestamp (in ms), ip_address, user, action, domain, campaign_id, cost, session_id

## timestamp converstions testing site : http://www.epochconverter.com/


## ----- config
days=4
entries_per_day=2000000
# file size calculations
#  format     entries_per_day     file_size in MB
#  csv       10000000             830 M file
#  json      1000000              169 M
#  json      3200000              500 M

#log_format="csv"
log_format="json"
## --- end config


import os
import datetime as dt
import random
import json

domains = ['facebook.com',  'yahoo.com',   'google.com',   'zynga.com',    'wikipedia.org',   'sf.craigslist.org',   'twitter.com',    'amazon.com',    'flickr.com',    'cnn.com',      'usatoday.com',      'npr.org',    'foxnews.com',    'comedycentral.com',   'youtube.com',   'hulu.com',   'bbc.co.uk',  'nytimes.com',   'sfgate.com',   'funnyordie.com']

actions = ['viewed', 'clicked', 'blocked']
total_users = 100000
total_ips = 1000
total_sessions = 100
total_campaigns = 20

# overwrite this function to customize log generation
def generate_log(timestamp):
  user = "user_%d" % random.randint(1,total_users)
  action = random.choice(actions)
  domain = random.choice(domains)
  ip_address = "ip_%d" % random.randint(1,total_ips)
  campaign = "campaign_%d" % random.randint(1,total_campaigns)
  session = "session_%s" % random.randint(1,total_sessions)
  #cost is in cents, could be zero
  cost = random.randint(1,200) - 20
  if (cost < 0):
    cost = 0

  #csv
  if (log_format == 'csv'):
    logline = "%s,%s,%s,%s,%s,%s,%s,%s" % (timestamp, ip_address, user, action, domain, campaign, cost, session)

  # generate JSON format
  if (log_format == 'json'):
    dict={'timestamp': timestamp, 'ip': ip_address, 'user': user, 'action': action, 'domain': domain,  'campaign':campaign, 'cost': cost,  'session': session}
    logline = json.dumps(dict)


  #print logline
  return logline



#main
## --- script main
if __name__ == '__main__':

  # mkdir
  dir="json"
  os.mkdir(dir)

  time_inc_ms = int ((24.0*3600*1000)/entries_per_day)
  #print "time inc ms", time_inc_ms
  #epoch = dt.datetime.fromtimestamp(0)
  epoch = dt.datetime(1970,1,1)

  year_start = dt.datetime(2015, 1, 1)
  for day in range(0, days):
    day_delta = dt.timedelta(days=day)
    start_ts = year_start + day_delta
    #end_ts = dt.datetime(start_ts.year, start_ts.month, start_ts.day, 23, 59, 59)
    end_ts = dt.datetime(start_ts.year, start_ts.month, start_ts.day+1, 0, 0, 0)
    filename = start_ts.strftime("%Y-%m-%d") 
    if (log_format == 'csv'):
      filename = filename + ".csv"
    if (log_format == 'json'):
      filename = filename + ".json"

    filename = dir + "/" + filename

    #print start_ts
    #print end_ts
    last_ts = start_ts

    with open(filename, "w") as fout:
      print "generating log ", filename
      while (last_ts < end_ts):
        delta_since_epoch = last_ts - epoch
        millis = int((delta_since_epoch.microseconds + (delta_since_epoch.seconds + delta_since_epoch.days * 24 * 3600) * 10**6) / 1e3)
        #print "last ts", last_ts
        #print "millis",  millis
        logline = generate_log(millis)
        fout.write(logline + "\n")

        last_ts = last_ts + dt.timedelta(milliseconds=time_inc_ms)

